core/src/main/scala/org/apache/spark/util/ListenerBus.scala

```diff
@@ -17,53 +17,222 @@
 
 package org.apache.spark.util
 
-import java.util.concurrent.CopyOnWriteArrayList
+import java.util.concurrent._
+import java.util.concurrent.atomic.{AtomicBoolean, AtomicLong}
+import java.util.concurrent.locks.{Condition, ReentrantLock}
 
 import scala.collection.JavaConverters._
 import scala.reflect.ClassTag
 import scala.util.control.NonFatal
 
+import com.google.common.util.concurrent.ThreadFactoryBuilder
+
 import org.apache.spark.internal.Logging
+import org.apache.spark.scheduler.LiveListenerBus
+
+private class ListenerEventExecutor[L <: AnyRef] (listenerName: String, queueCapacity: Int)
```

Comment:
Not a fan of the name, since you don't execute events. I prefer `ListenerEventDispatcher.`

Also, `listenerName` is more of a group name now, right? You mention "listener" in many places of this class, those should be replaced with "listener group".

Finally, not sure why do you need the type parameter?

nit: no space before `(`